### ISSUE TITLE
What needs to happen before API is feature-complete

### ISSUE THREAD

**Issue by RonnChyran** [SnowflakePowered/snowflake#12]
Emulator API
- [x] A way to store and load custom game profiles during configuration compilation (`ConfigurationStore`)
- [x]  A way to store and load configured controller profiles (filesystem?)
- [x]  A way to store and load emulator-specific flags for certain games (use a database)
- [x]  Implementing common logic in the abstract base class

Others
- [x] A massive refactoring is required, sorting every class into a distinct API
- [x] Documenting interfaces for every core class
- [ ]  Completely remove the SharpYaml dependency, just migrate everything to json.net.
- [x]  EventManager API should be implemented AFTER refactoring
- [x]  Unit and integration tests if possible

**Comment by RonnChyran** [comment]
As we have the frameworks for tests and Events ready I'll be closing this and splitting off the SharpYaml dependency as a separate issue.
